### DISCUSSION TITLE
Recommended way to link to PDF outputs from a MyST website?

### DISCUSSION THREAD

**Discussion by choldgraf** [jupyter-book/mystmd#1016] [Q&A] [answered]
I'd like to be able to build a MyST website that includes a link to a PDF version of the whole site (so that people can download it locally). Right now the only way I can imagine doing this is to run the MyST build twice, and in the HTML theme (where?) manually point to the PDF output so users can download it.

Is there a more graceful way to build a PDF + HTML site with MyST and make the PDF available via the MyST site?

**Comment by rowanc1** [comment]
I have run into similar challenges. Current practice is build the exports, then link them in as you suggest, for example, with a static link in the markdown that turns into a download, or in the site actions in the nav (add `static: true`). In #1009 I have started to think we need a `downloads` section for a site (or page) which does a better job at exposing the downloads we want to show to a user (e.g. the PDF, JATS, Meca, source, etc.). Hadn't thought about the build order, but that also makes sense to be part of the thinking, we have precedence for that in building the meca bundle (a zip file used in publishing), which needs to wait until the other exports are built.

**Reply by choldgraf** [reply]
Could you explain what you mean by this?

> in the site actions in the nav (add static: true)

I did various searches for `site actions`, `static: true` etc in the docs and don't see anything about this, so not sure what to put.

**Reply by rowanc1** [reply]
Site actions are [briefly described here](https://mystmd.org/guide/quickstart-myst-websites#configuration). If the action is a PDF, you need to add a `static: true` currently (this is not documented at the moment, I think that removing that requirement may be just as easy as adding docs actually, either way issues welcome! 🐛📖).

**Reply by choldgraf** [reply]
OK I got really close to getting this to work. You can find my GitHub Action for this here:

https://github.com/2i2c-org/report-czi-2021/blob/bfa9c87e8c4087873d6fab9b5c3bc84b5db802b6/.github/workflows/deploy.yml#L36-L48

The only problem is that there's an error after the report PDF file is generated...I can't tell but it seems like it might be a github actions thing? Here's the error message:

<details>
<summary>Error message</summary>

```
Run myst build --typst
📬 Performing exports:
   myst.yml -> _build/pdf/report.pdf
📖 Built year1.md in 110 ms.
🪄  Linked 2 DOIs in 69[7](https://github.com/2i2c-org/report-czi-2021/actions/runs/8369638654/job/22915713776#step:6:8) ms for main.md
📖 Built main.md in 1.03 s.
🪄  Linked [8](https://github.com/2i2c-org/report-czi-2021/actions/runs/8369638654/job/22915713776#step:6:9) DOIs in 1.01 s for year3.md
📖 Built year3.md in 1.24 s.
🪄  Linked 7 DOIs in 1.18 s for year2.md
📖 Built year2.md in 1.38 s.
📚 Built 4 pages for export (including 0 dependencies) from /home/runner/work/report-czi-2021/report-czi-2021 in 1.46 s.
🐕 Fetching template from https://github.com/rowanc1/typst-book/archive/refs/heads/main.zip
💾 Saved template to path _build/templates/typst/a7f4d4fcb0[9](https://github.com/2i2c-org/report-czi-2021/actions/runs/8369638654/job/22915713776#step:6:10)d7d9ef0d3fc42445a9263fb8de44a54644c0975db0d0d4775965e
📑 Exported typst in 727 ms, copying to _build/temp/mystg2IwFo/report.typ
🖨 Rendering typst pdf to _build/pdf/report.pdf
Error: ENOENT: no such file or directory, copyfile '/home/runner/work/report-czi-[20](https://github.com/2i2c-org/report-czi-2021/actions/runs/8369638654/job/22915713776#step:6:21)[21](https://github.com/2i2c-org/report-czi-2021/actions/runs/8369638654/job/22915713776#step:6:22)/report-czi-2021/_build/temp/mystg2IwFo/report.pdf' -> '/home/runner/work/report-czi-2021/report-czi-2021/_build/pdf/report.pdf'
```

</details>

I sent you an invite in case you wanna look at the logs, in case this is somehow uncovering a bug that only happens on github actions or something

https://github.com/2i2c-org/report-czi-2021/invitations

**Reply by choldgraf** [reply]
I am wondering if there's a clue in this error message. If I add the logs from _successful_ copy events, you see something like:

```
📑 Exported typst in 716 ms, copying to _build/temp/mystXXrVnO/report.typ
typst CLI must be installed to build PDFs from typst
🖨 Rendering typst pdf to _build/pdf/report.pdf

Error: ENOENT: no such file or directory, copyfile '/home/runner/work/report-czi-2021/report-czi-2021/_build/temp/mystXXrVnO/report.pdf' -> '/home/runner/work/report-czi-2021/report-czi-2021/_build/pdf/report.pdf'
```

It seems like there's something going wrong in how the temporary file (`temp/mystXXrVnO/report.pdf`) is created in github actions?

**Reply by rowanc1** [reply, marked as answer]
You should install typst, this is the easiest way:

```
- name: Install Typst for PDF builds
  uses: typst-community/setup-typst@v3
```

The PDF did not get built, but tried to copy. The logs are certainly misleading though.

**Reply by choldgraf** [reply]
amazing, that solved it! check out my nifty PDF build at https://2i2c.org/report-czi-2021

![CleanShot 2024-03-21 at 15 33 08@2x](https://github.com/executablebooks/mystmd/assets/1839645/1d695eb9-9da5-48e3-936f-af347708f5e1)
